File: mns-reactor/src/main/java/cz/fb/manaus/reactor/charge/ChargeGrowthForecaster.java
package cz.fb.manaus.reactor.charge;

import com.google.common.collect.LinkedListMultimap;
import cz.fb.manaus.core.model.Bet;
import cz.fb.manaus.core.model.MarketSnapshot;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.provider.ExchangeProvider;
import cz.fb.manaus.reactor.betting.AmountAdviser;
import cz.fb.manaus.reactor.price.Fairness;
import cz.fb.manaus.reactor.price.ProbabilityCalculator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.OptionalDouble;

@Component
public class ChargeGrowthForecaster {

    @Autowired
    private MarketChargeSimulator simulator;
    @Autowired
    private ProbabilityCalculator probabilityCalculator;
    @Autowired
    private AmountAdviser amountAdviser;
    @Autowired
    private ExchangeProvider exchangeProvider;


    private LinkedListMultimap<Long, Price> convertBetData(List<Bet> currentBets) {
        var bets = LinkedListMultimap.<Long, Price>create();
        for (var bet : currentBets) {
            var price = bet.getRequestedPrice().getPrice();
            var matchedAmount = bet.getMatchedAmount();
            var side = bet.getRequestedPrice().getSide();
            bets.put(bet.getSelectionId(), new Price(price, matchedAmount, side));
        }
        return bets;
    }

    public OptionalDouble getForecast(long selectionId, Side betSide,
                                      MarketSnapshot snapshot, Fairness fairness) {
        if (exchangeProvider.isPerMarketCharge()) {
            var fairnessSide = fairness.getMoreCredibleSide();
            if (fairnessSide.isPresent()) {
                var sideFairness = fairness.get(fairnessSide.get());
                var probabilities = probabilityCalculator.fromFairness(
                        sideFairness.getAsDouble(), fairnessSide.get(), snapshot.getMarketPrices());
                var marketPrices = snapshot.getMarketPrices();
                var bets = convertBetData(snapshot.getCurrentBets());
                var runnerPrices = marketPrices.getRunnerPrices(selectionId);
                var winnerCount = marketPrices.getWinnerCount();
                var oldCharge = simulator.getChargeMean(winnerCount, exchangeProvider.getChargeRate(), probabilities, bets);
                var bestPrice = runnerPrices.getHomogeneous(betSide.getOpposite()).getBestPrice();
                if (bestPrice.isPresent()) {
                    var price = bestPrice.get().getPrice();
                    var amount = amountAdviser.getAmount();
                    bets.put(selectionId, new Price(price, amount, betSide));
                    var newCharge = simulator.getChargeMean(winnerCount, exchangeProvider.getChargeRate(), probabilities, bets);
                    var result = newCharge / oldCharge;
                    return OptionalDouble.of(result);
                }
            }
        }
        return OptionalDouble.empty();
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/charge/ChargeGrowthForecaster.kt
package cz.fb.manaus.reactor.charge

import cz.fb.manaus.core.model.MarketSnapshot
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.SideSelection
import cz.fb.manaus.core.model.TrackedBet
import cz.fb.manaus.reactor.betting.AmountAdviser
import cz.fb.manaus.reactor.price.Fairness
import cz.fb.manaus.reactor.price.ProbabilityCalculator
import cz.fb.manaus.reactor.price.getRunnerPrices
import org.springframework.stereotype.Component
import kotlin.math.min

@Component
class ChargeGrowthForecaster(
        private val simulator: MarketChargeSimulator,
        private val amountAdviser: AmountAdviser
) {

    private fun convertBetData(currentBets: List<TrackedBet>): Map<Long, List<Price>> {
        return currentBets.groupBy({ it.remote.selectionId }, {
            val bet = it.remote
            val price = bet.requestedPrice.price
            val matchedAmount = bet.matchedAmount
            val side = bet.requestedPrice.side
            Price(price, matchedAmount, side)
        })
    }

    fun getForecast(
            sideSelection: SideSelection,
            snapshot: MarketSnapshot,
            fairness: Fairness,
            commission: Double
    ): Double? {
        val (side, selectionId) = sideSelection
        val fairnessSide = fairness.moreCredibleSide
        if (fairnessSide != null) {
            val sideFairness = fairness[fairnessSide]!!
            val probabilities = ProbabilityCalculator.fromFairness(
                    sideFairness, fairnessSide, snapshot.runnerPrices
            )
            val marketPrices = snapshot.runnerPrices
            val bets = convertBetData(snapshot.currentBets).toMutableMap()
            val runnerPrices = getRunnerPrices(marketPrices, selectionId)
            val oldCharge = simulator.getChargeMean(
                    winnerCount = 1,
                    commission = commission,
                    probabilities = probabilities,
                    bets = bets
            )
            val bestPrice = runnerPrices.getHomogeneous(side.opposite).bestPrice
            if (bestPrice != null) {
                val price = bestPrice.price
                val amount = amountAdviser.amount
                val selBets = bets[selectionId] ?: emptyList()
                bets[selectionId] = selBets + listOf(Price(price, amount, side))
                val newCharge = simulator.getChargeMean(
                        winnerCount = 1,
                        commission = commission,
                        probabilities = probabilities,
                        bets = bets
                )
                return min(1000.0, newCharge / oldCharge)
            }
        }
        return null
    }

}
